### DISCUSSION TITLE
Can solver::unsat_core() be interrupted by context::interrupt()?

### DISCUSSION THREAD

**Discussion by m-carrasco** [Z3Prover/z3#6310] [Q&A]
Hi,

I just wanted to ask if a solver computing `unsat_core()` can be interrupted by [`context::interrupt()`](https://z3prover.github.io/api/html/classz3_1_1context.html#aebc9afd317d656680550a13818e71917). In other words, I'd like to know if `solver::unsat_core()` is interruptable. In this case, the solver is configured to minimise the unsat core.

Thanks in advance!

Best regards,
Manuel

**Comment by m-carrasco** [comment]
In case it helps anybody, `context::interrupt()` can interrupt `unsat_core()`. When `unsat_core()` is interrupted, it looks like it returns the best result found so far. 

However, I am not sure if the interruption based on a timeout can be implemented using Z3 API. At the moment, I had to work around it using a different thread.

**Reply by NikolajBjorner** [reply]
I added handlers for timeout inside the API for unsat cores. It should now respect timeouts and resource limits

**Reply by m-carrasco** [reply]
Thank you so much! I will test it as soon as I can.

**Reply by m-carrasco** [reply]
@NikolajBjorner Just to learn, why is an exception thrown when the [unsat core is empty](https://github.com/Z3Prover/z3/commit/ddf4895c2f4cab93216233fb10f600158b41f6bf#diff-a460cdb07afec6af7317bf9c3b93d762c0d16cd839f59bf67a210408259d184eR713)?

**Reply by NikolajBjorner** [reply]
it isn't safe: if it ends up canceling before extracting the first core it will be empty because it wasn't filled in. This is unsound. 
So choosing between unsoundness and inconvenience, I re-throw on timeout/exception and empty core.
